File: w3/dz_java/src/Basic/Point/Point.java
import static java.lang.Math.hypot;

public class Point {
    private double x;
    private double y;

    public Point() {
        this.x = 0;
        this.y = 0;
    }

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public void setX(double x) {
        this.x = x;
    }

    public void setY(double y) {
        this.y = y;
    }

    public double distance(Point other) {
        return hypot(this.x - other.x, this.y - other.y);
    }

    public boolean equal(Point other) {
        return this.x == other.x && this.y == other.y;
    }

    public boolean notEqual(Point other) {
        return !this.equal(other);
    }

    public String toString() {
        return String.format( "(" + this.getX() + ", " + this.getY() + ")");
    }
}

File: w4/php_dz/Car/Point.php
<?php
    class Point {
        private $x, $y;

        private function validate($value) {
            if ( is_numeric($value) ) {
                return $value;
            }
            throw new Exception("Invalid param");
        }

        public function __construct($x = 0, $y = 0) {
            $this->x = $this->validate($x);
            $this->y = $this->validate($y);
        }

        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw Exception("Attribute error: attribute $name not found");
        }

        public function __set($name, $value) {
            if ( property_exists($this, $name) ) {
                $this->$name = $this->validate($value);
            } else {
                throw Exception("Attribute error: attribute $name not found");
            }
        }

        public function distance(Point &$other) {
            return hypot($this->x - $other->x, $this->y - $other->y);
        }

        public function compare(Point &$other) {
            if ( $this == $other ) {
                echo $this . " is equal to " . $other .PHP_EOL;
            } else {
                echo $this . " is not equal to " . $other .PHP_EOL;
            }
        }

        public function __toString() {
            return sprintf("(%d, %d)", $this->x, $this->y);
        }
    }
?>